python/TestHarness/schedulers/Job.py

```diff
@@ -82,6 +95,42 @@ def __init__(self, tester, job_dag, options):
         # Initialize jobs with a holding status
         self.setStatus(self.hold)
 
+    def getUpstream(self, a_job):
+        """ Method to return a list of all the jobs that need to be run before this job """
+        somejobs = self.getDAG()
+        t_list = []
+        for temp_job in somejobs.predecessors(a_job):
+            t_list.append(temp_job.getTestName())
+        return t_list
+
+    def getDownstreams(self, a_job, graph):
+        """ Method to return a list of all the jobs that need to be run after this job """
+        somejobs = self.getDAG()
+        t_list = []
+        for temp_job in somejobs.all_downstreams(a_job):
+            t_list.append(temp_job.getTestName())
+        return t_list
```

Comment:
I am little confused as to the purpose of DAG node management within the Job class. Is there a way we can move DAG operations into the JobDAG instead? Basically, I'm asking you if it is possible to instead create wrapper methods into the JobDAG object instead of doing the actual JobDAG work in the Job object. Something like:

JobDAG.py:
```python
def getUpstream(self, job):
    """ Method to return a list of all the jobs that need to be run before this job """
    t_list = []
        for temp_job in self.__job_dag.predecessors(job):
            t_list.append(temp_job.getTestName())
        return t_list

def getDownstreams(self, job):
    """ Method to return a list of all the jobs that need to be run after this job """
    t_list = []
    for temp_job in self.__job_dag.all_downstreams(job):
        t_list.append(temp_job.getTestName())
    return t_list
```

Job.py then looks like:
```python
def getUpstream(self):
    return self.__job_dag.getUpstream(self)

def getDownstreams(self):
    return self.__job_dag.getUpstream(self)
```

Reply:
Also... Cody and I have repeatedly not liked each other's naming convention for upstream/dependents/predecessors, etc. We really should take a vote on what to call this feature. getUpstream(C) does what exact?

A <-- B <-- C (as the DAG constructs it based on prereqs)

returns A and B?

But when we write the DAG on a white board like this:

A --> B --> C (a runs, then b runs, then c...)

what then should 'getUpstream' return? -this is where the confusion lies. To re-iterate: You are not doing anything incorrect. I would just like to call the vote on what to call these two procedures.

Reply:
Ok, I think anything referring to getting information about the JobDAG should be in the JobDAG, so these functions would be better in the JobDAG class with wrappers in the Job Class.  I was thinking though that each job itself has a unique set of predecessors, descendants, and race partners , so I was storing that unique set or list in the Job Class.  Is that alright?  In addition, when I tried to store the predecessor and descendant variable information in a DAG in the JobDAG, I found that they were getting deleted by the JobDAG's self-cleaning.

Reply:
I think storing this information in the Job class is a good idea. But performing the work... yeah that belongs elsewhere (in the JobDAG class).

Yeah, I got bitten by that as well one day. Is why I created getOriginalDAG in (in the dag object). While I peruse over that method, it is hacky. It is very difficult to understand when a DAG is finished, with no more incoming changes. In fact, my PR #13424 addresses this by removing as many places where it gets re-set as possible.